### ISSUE TITLE
Phase 39.3 — CounterfactualGenerator: DiCE, Contrastive Explanations & Actionable Recourse

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#790]
## Phase 39.3 — CounterfactualGenerator

> Counterfactual Explanations, DiCE, Contrastive Reasoning, Actionable Recourse

### Overview

The `CounterfactualGenerator` answers "what would need to change?" questions by generating minimal input perturbations that flip a model's prediction. Implements Wachter et al. (2017) gradient-based counterfactuals, Mothilal et al. (2020) DiCE for diverse counterfactual sets, and actionable recourse constraints ensuring suggestions are feasible.

### Architecture

```
┌──────────────────────────────────────────────┐
│          CounterfactualGenerator             │
│  ┌──────────┐  ┌──────────┐  ┌────────────┐ │
│  │ Wachter  │  │  DiCE    │  │ Recourse   │ │
│  │ Method   │  │  Engine  │  │ Validator  │ │
│  └────┬─────┘  └────┬─────┘  └─────┬──────┘ │
│       │              │              │        │
│  ┌────▼──────────────▼──────────────▼──────┐ │
│  │     Feasibility Constraint Engine       │ │
│  │  (immutable features, causal graph,     │ │
│  │   monotonicity, range constraints)      │ │
│  └────────────────┬────────────────────────┘ │
│                   │                          │
│  ┌────────────────▼────────────────────────┐ │
│  │    Contrastive Explanation Formatter     │ │
│  │  ("Because X, not Y" explanations)      │ │
│  └─────────────────────────────────────────┘ │
└──────────────────────────────────────────────┘
```

### Frozen Dataclasses

```python
@dataclass(frozen=True)
class CounterfactualConfig:
    """Configuration for counterfactual generation."""
    n_counterfactuals: int = 5               # number of diverse CFs to generate
    proximity_weight: float = 0.5            # λ₁: distance to original
    diversity_weight: float = 1.0            # λ₂: diversity among CFs
    sparsity_weight: float = 0.1             # λ₃: number of features changed
    immutable_features: Tuple[int, ...] = () # features that cannot be changed
    feature_ranges: Tuple[Tuple[float, float], ...] = ()  # (min, max) per feature
    causal_constraints: Tuple[Tuple[int, int], ...] = ()  # (cause, effect) pairs

@dataclass(frozen=True)
class Counterfactual:
    """A single counterfactual explanation."""
    original: Tuple[float, ...]              # original input
    counterfactual: Tuple[float, ...]        # modified input
    original_prediction: float               # model output for original
    cf_prediction: float                     # model output for counterfactual
    features_changed: Tuple[int, ...]        # indices of changed features
    distance: float                          # L1 distance to original
    validity: bool                           # does CF achieve target prediction?

@dataclass(frozen=True)
class RecourseAction:
    """Actionable recourse recommendation."""
    feature_index: int
    feature_name: str
    current_value: float
    target_value: float
    action_description: str                  # human-readable action
    cost: float = 0.0                        # estimated action cost
    feasibility: float = 1.0                 # feasibility score [0, 1]
```

### Protocol Interface

```python
@runtime_checkable
class CounterfactualGeneratorProtocol(Protocol):
    """Protocol for counterfactual explanation generators."""

    def generate(
        self,
        model: Any,
        instance: np.ndarray,
        target_class: int,
        config: CounterfactualConfig,
    ) -> list[Counterfactual]:
        """Generate diverse counterfactual explanations."""
        ...

    def compute_recourse(
        self,
        model: Any,
        instance: np.ndarray,
        target_class: int,
        config: CounterfactualConfig,
    ) -> list[RecourseAction]:
        """Compute actionable recourse recommendations."""
        ...

    def contrastive_explain(
        self,
        model: Any,
        fact: np.ndarray,
        foil: np.ndarray,
    ) -> str:
        """Generate contrastive explanation: why fact, not foil?"""
        ...

    def validate_counterfactual(
        self,
        cf: Counterfactual,
        config: CounterfactualConfig,
    ) -> bool:
        """Check counterfactual satisfies all constraints."""
        ...
```

### Acceptance Criteria

- [ ] Wachter et al. gradient-based CF with λ-weighted loss (proximity + validity)
- [ ] DiCE with diversity-promoting DPP-based sampling
- [ ] Immutable feature constraints (age, race cannot be changed)
- [ ] Causal monotonicity constraints via directed acyclic graph
- [ ] Actionable recourse with cost estimation and feasibility scoring
- [ ] Contrastive "because X, not Y" natural language explanations
- [ ] Batch generation with configurable diversity-proximity tradeoff
- [ ] Integration with Phase 23 DecisionTreeSolver for decision-path CFs

### References

- Wachter et al. (2017) — Counterfactual Explanations without Opening the Black Box
- Mothilal et al. (2020) — DiCE: Diverse Counterfactual Explanations
- Ustun et al. (2019) — Actionable Recourse in Linear Classification
- Miller (2019) — Explanation in Artificial Intelligence: Insights from Social Sciences
- Karimi et al. (2020) — Algorithmic Recourse: from Counterfactual Explanations to Interventions

**Comment by web3guru888** [comment]
## Implementation Notes — CounterfactualGenerator Code Skeleton

```python
"""asi_build/xai/counterfactual_generator.py — Phase 39.3"""
from __future__ import annotations

import numpy as np
from dataclasses import dataclass
from typing import Any

@dataclass(frozen=True)
class CounterfactualConfig:
    n_counterfactuals: int = 5
    proximity_weight: float = 0.5
    diversity_weight: float = 1.0
    sparsity_weight: float = 0.1
    immutable_features: tuple[int, ...] = ()
    feature_ranges: tuple[tuple[float, float], ...] = ()
    causal_constraints: tuple[tuple[int, int], ...] = ()

@dataclass(frozen=True)
class Counterfactual:
    original: tuple[float, ...]
    counterfactual: tuple[float, ...]
    original_prediction: float
    cf_prediction: float
    features_changed: tuple[int, ...]
    distance: float
    validity: bool

@dataclass(frozen=True)
class RecourseAction:
    feature_index: int
    feature_name: str
    current_value: float
    target_value: float
    action_description: str
    cost: float = 0.0
    feasibility: float = 1.0


class CounterfactualGenerator:
    """Diverse counterfactual explanation generator (Wachter 2017, Mothilal 2020)."""

    def __init__(self, max_iterations: int = 500, learning_rate: float = 0.01):
        self._max_iter = max_iterations
        self._lr = learning_rate

    def generate(self, model: Any, instance: np.ndarray,
                 target_class: int, config: CounterfactualConfig) -> list[Counterfactual]:
        """Generate diverse counterfactuals using DiCE algorithm."""
        cfs = []
        # Initialize k counterfactual candidates near the instance
        candidates = [instance + np.random.normal(0, 0.1, size=instance.shape)
                      for _ in range(config.n_counterfactuals)]

        for iteration in range(self._max_iter):
            for i, cf in enumerate(candidates):
                # Compute combined loss
                pred = model.predict_proba(cf.reshape(1, -1))[0][target_class]
                validity_loss = max(0, 0.5 - pred)  # hinge loss
                proximity_loss = np.sum(np.abs(cf - instance))  # L1 distance
                # Diversity: distance to other CFs
                diversity = sum(np.sum(np.abs(cf - candidates[j]))
                                for j in range(len(candidates)) if j != i)
                # Sparsity: number of changed features
                sparsity = np.sum(np.abs(cf - instance) > 1e-6)

                loss = (config.proximity_weight * proximity_loss
                        - config.diversity_weight * diversity / max(1, len(candidates) - 1)
                        + config.sparsity_weight * sparsity
                        + 10.0 * validity_loss)

                # Gradient step (numerical gradient)
                grad = self._numerical_gradient(
                    model, instance, cf, target_class, config, candidates, i)
                cf = cf - self._lr * grad

                # Enforce immutable features
                for idx in config.immutable_features:
                    cf[idx] = instance[idx]

                # Enforce feature ranges
                if config.feature_ranges:
                    for idx, (lo, hi) in enumerate(config.feature_ranges):
                        cf[idx] = np.clip(cf[idx], lo, hi)

                candidates[i] = cf

        # Convert to Counterfactual objects
        for cf in candidates:
            pred_orig = float(model.predict(instance.reshape(1, -1))[0])
            pred_cf = float(model.predict(cf.reshape(1, -1))[0])
            changed = tuple(int(i) for i in range(len(instance)) if abs(cf[i] - instance[i]) > 1e-6)
            dist = float(np.sum(np.abs(cf - instance)))
            valid = model.predict(cf.reshape(1, -1))[0] == target_class

            cfs.append(Counterfactual(
                original=tuple(instance.tolist()),
                counterfactual=tuple(cf.tolist()),
                original_prediction=pred_orig,
                cf_prediction=pred_cf,
                features_changed=changed,
                distance=dist,
                validity=bool(valid),
            ))

        return sorted(cfs, key=lambda c: c.distance)

    def compute_recourse(self, model: Any, instance: np.ndarray,
                         target_class: int, config: CounterfactualConfig,
                         feature_names: list[str] | None = None) -> list[RecourseAction]:
        """Compute actionable recourse from best counterfactual."""
        cfs = self.generate(model, instance, target_class, config)
        best_cf = next((c for c in cfs if c.validity), cfs[0])
        actions = []
        for idx in best_cf.features_changed:
            name = feature_names[idx] if feature_names else f"feature_{idx}"
            delta = best_cf.counterfactual[idx] - best_cf.original[idx]
            direction = "increase" if delta > 0 else "decrease"
            actions.append(RecourseAction(
                feature_index=idx,
                feature_name=name,
                current_value=best_cf.original[idx],
                target_value=best_cf.counterfactual[idx],
                action_description=f"{direction} {name} from {best_cf.original[idx]:.2f} to {best_cf.counterfactual[idx]:.2f}",
                cost=abs(delta),
                feasibility=1.0 if idx not in config.immutable_features else 0.0,
            ))
        return sorted(actions, key=lambda a: a.cost)

    def contrastive_explain(self, model: Any, fact: np.ndarray, foil: np.ndarray,
                            feature_names: list[str] | None = None) -> str:
        """Generate contrastive explanation: why fact, not foil?"""
        diff_indices = [i for i in range(len(fact)) if abs(fact[i] - foil[i]) > 1e-6]
        parts = []
        for idx in diff_indices:
            name = feature_names[idx] if feature_names else f"feature_{idx}"
            parts.append(f"{name} is {fact[idx]:.2f} vs {foil[idx]:.2f}")
        pred_fact = model.predict(fact.reshape(1, -1))[0]
        pred_foil = model.predict(foil.reshape(1, -1))[0]
        return f"Predicted {pred_fact} instead of {pred_foil} because: {'; '.join(parts)}"

    def _numerical_gradient(self, model, instance, cf, target_class, config, candidates, idx):
        """Numerical gradient of the DiCE loss."""
        eps = 1e-4
        grad = np.zeros_like(cf)
        for i in range(len(cf)):
            if i in config.immutable_features:
                continue
            cf_plus, cf_minus = cf.copy(), cf.copy()
            cf_plus[i] += eps
            cf_minus[i] -= eps
            loss_plus = self._dice_loss(model, instance, cf_plus, target_class, config, candidates, idx)
            loss_minus = self._dice_loss(model, instance, cf_minus, target_class, config, candidates, idx)
            grad[i] = (loss_plus - loss_minus) / (2 * eps)
        return grad

    def _dice_loss(self, model, instance, cf, target_class, config, candidates, idx):
        pred = model.predict_proba(cf.reshape(1, -1))[0][target_class]
        validity = max(0, 0.5 - pred)
        proximity = np.sum(np.abs(cf - instance))
        diversity = sum(np.sum(np.abs(cf - candidates[j])) for j in range(len(candidates)) if j != idx)
        sparsity = np.sum(np.abs(cf - instance) > 1e-6)
        return config.proximity_weight * proximity - config.diversity_weight * diversity / max(1, len(candidates)-1) + config.sparsity_weight * sparsity + 10 * validity
```

### mypy type-check table

| Symbol | Type | Status |
|--------|------|--------|
| `CounterfactualGenerator.generate` | `(Any, ndarray, int, CFConfig) -> list[Counterfactual]` | ✅ |
| `CounterfactualGenerator.compute_recourse` | `(Any, ndarray, int, CFConfig) -> list[RecourseAction]` | ✅ |
| `CounterfactualGenerator.contrastive_explain` | `(Any, ndarray, ndarray) -> str` | ✅ |
